### ISSUE TITLE
[3.3.5] Core/Spells: crash while processing spell targets

### ISSUE THREAD

**Issue by calaluis** [TrinityCore/TrinityCore#22158]
**Description:**

When entering an instance to later mock enemies, the emulator gives an internal error.

**Current behaviour:**

CHANGEME Tell us what happens.
https://gist.github.com/calaluis/9ca779ee259ee1c44d79975ccd3dfb7d

**Expected behaviour:**

I should not fall and function properly to be able to finish the dungeon.

**Steps to reproduce the problem:**

1. Enter instance.
2. Mob enemies.
3. Launch skills

**Branch(es):**

3.3.5

**TC rev. hash/commit:** 

https://github.com/TrinityCore/TrinityCore/commit/2ef33158486831aa061050f5cc7d316fee17721d

**Operating system:** 

Windows Server 2012 R2, full updated.

Technical screen capture.

![image](https://user-images.githubusercontent.com/33674615/42729589-a45e43de-87aa-11e8-9335-fafe32d84342.png)

**Comment by TrinityCoreBot** [comment]
This issue is not considered valid because of the following reasons:
- No proper commit hash (revision)

Please read http://bit.ly/tc-issuetracker-and-you and add the missing information to **this issue**. Thanks.

**Comment by Aokromes** [comment]
any specific class/dungeon/race pairs?

**Comment by calaluis** [comment]
Sorry for not providing more details, as this error had reported in the forum a few weeks ago without getting a response.

Below I detail the way to cause this error.

First: A group of two people was made to go to the Shadowfang Keep dungeon. The first character is a protection paladin and the second is a shadow priest. Both are Blood Elf.

Second: When trying to gather several enemies and launch the skills, the emulator of the worldserver.exe file triggered that error.

I leave the URL of the post I left in the forum, which is the same but I have reported it there as well.

https://community.trinitycore.org/topic/14211-emulator-worldserver-crashes/

I also leave a ZIP file containing the DMP file that the emulator generates when it falls.

[ErrWoW.zip](https://github.com/TrinityCore/TrinityCore/files/2195462/ErrWoW.zip)

Stay tuned.

**Comment by jackpoz** [comment]
Spell ID 31789 http://www.wowhead.com/spell=31789/righteous-defense

btw do you have any custom change at all ?

**Comment by calaluis** [comment]
Hello jackpoz,

The skill was not always used.

http://www.wowhead.com/spell=31789/righteous-defense

The one that was always used is this one.

http://bfa.wowhead.com/spell=26573/consecration

El paladin es de nivel 31, y el sacerdote es de nivel 25.

There is nothing personalized in the emulator, nor in the core; and I can even tell you that I downloaded the last revision of branch 3.3.5 to apply CMAKE, then compile with Visual Studio (everything from scratch).

(Excuse me if English is very bad, since I am using Google Translator to be able to communicate).

Thanks illfated for the suggestion.

**Comment by calaluis** [comment]
Dear,

I share this video that I uploaded to youtube that shows the error that the emuador has.

Link: https://youtu.be/Yz9OSqJw77o

**Comment by calaluis** [comment]
Hello again,

Inquiring more into the subject, the error that I have reported will not be related to this other incident in the following link:

https://github.com/TrinityCore/TrinityCore/issues/21765

**Comment by Treeston** [comment]
Alright, so some dump analysis first:

* `target` (=`<begin>$L1`, at `0xE89DB6CD10`) is inaccessible memory
* The `Spell` object (`this` at `0xE8EE03E9D0`) itself _appears_ to be fine at a glance, though this is hard to evaluate
* The parameter `container` (`0xE8EE03EBF0`, this is equal to `m_UniqueTargetInfo` of the `Spell`) doesn't appear to be corrupted

**Comment by Treeston** [comment]
Container invalidations like these are always a pain to deal with. Guesswork and a healthy dose of luck, at best.

So I did some digging. Righteous Defense uses `RandomResize`, which caught my interest, since it's a fairly rarely-used (and also fairly recent) implementation.

In there, we `std::move` between container iterators. That's fine. However, we also `std::move` between _possibly identical_ container iterators. Pedantically speaking, that leaves the referenced object in an indeterminate state.

64a61e03ab0799b5437ddb9fe702c4912f3f9dd9 adjusts this, so it's _possible_ that it was the underlying cause and this is now fixed. I really can't tell.

**Comment by Treeston** [comment]
@calaluis since you have video of this, can I assume that this is reliable for you?
Does it still happen on the latest rev. (64a61e0 or later)?
Does it happen with a new character too?
Does it happen in other zones, or in SFK only?
Any patterns that could help me reproduce this would be great.

**Comment by calaluis** [comment]
Hello @Treeston , the only thing I did to arrive at that error was the following:

1.- Download the emulator in the green "Clone or Download" button. Then I clicked on the "Download ZIP" option.
2.- All the steps of the documentation were executed to implement the elumador in a server, with Windows Server 2012 R2.
3.- What @jackpoz  says is absolutely right: You only use the Spell ID 31789 in party, no matter the map or the place where you are and triggers the error that I have reported in this locus. In party it is, because the spell only works giving protection to a member and not to himself.

Interesting the analysis that you have made, since I spent a whole day debugging the source code without being able to understand its functional process logic.

The video that I have uploaded, was made with the character of my brother (the paladin), and I the priest, to be able to recreate the error.

I hope that what you have said is useful.
